File: 0508/0508-144/src/history_manager.py
import numpy as np


class HistoryManager:
    def __init__(self, max_history=50):
        self.max_history = max_history
        self.undo_stack = []
        self.redo_stack = []
    
    def can_undo(self):
        return len(self.undo_stack) > 0
    
    def can_redo(self):
        return len(self.redo_stack) > 0
    
    def push(self, state):
        self.undo_stack.append(state.copy())
        if len(self.undo_stack) > self.max_history:
            self.undo_stack.pop(0)
        self.redo_stack.clear()
    
    def undo(self, current_state):
        if not self.can_undo():
            return None
        prev_state = self.undo_stack.pop()
        self.redo_stack.append(current_state.copy())
        return prev_state
    
    def redo(self, current_state):
        if not self.can_redo():
            return None
        next_state = self.redo_stack.pop()
        self.undo_stack.append(current_state.copy())
        return next_state
    
    def clear(self):
        self.undo_stack.clear()
        self.redo_stack.clear()
    
    def set_max_history(self, max_history):
        self.max_history = max_history
        while len(self.undo_stack) > self.max_history:
            self.undo_stack.pop(0)

File: 0508/0508-382/modules/HistoryManager.js
export class HistoryManager {
    constructor(maxHistory = 50) {
        this.history = [];
        this.currentIndex = -1;
        this.maxHistory = maxHistory;
    }

    push(state) {
        const clonedState = this.cloneImageData(state);
        
        this.history = this.history.slice(0, this.currentIndex + 1);
        this.history.push(clonedState);
        
        if (this.history.length > this.maxHistory) {
            this.history.shift();
        }
        
        this.currentIndex = this.history.length - 1;
    }

    undo() {
        if (this.currentIndex <= 0) {
            return null;
        }
        
        this.currentIndex--;
        return this.cloneImageData(this.history[this.currentIndex]);
    }

    redo() {
        if (this.currentIndex >= this.history.length - 1) {
            return null;
        }
        
        this.currentIndex++;
        return this.cloneImageData(this.history[this.currentIndex]);
    }

    canUndo() {
        return this.currentIndex > 0;
    }

    canRedo() {
        return this.currentIndex < this.history.length - 1;
    }

    clear() {
        this.history = [];
        this.currentIndex = -1;
    }

    cloneImageData(imageData) {
        const newData = new ImageData(imageData.width, imageData.height);
        newData.data.set(imageData.data);
        return newData;
    }
}